### PR TITLE
[FLINK-35346][table-common] Introduce workflow scheduler interface for materialized table

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Introduce workflow scheduler interface for materialized table*

## Brief change log

  - *Introduce workflow scheduler interface for materialized table*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests in WorkflowSchedulerFactoryUtilTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no )
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     * @throws WorkflowException if initializing workflow scheduler occur exception
+     */
+    void open() throws WorkflowException;
+
+    /**
+     * Close this workflow scheduler when it is no longer needed and release any resource that it
+     * might be holding.
+     *
+     * @throws WorkflowException if close the related resources of workflow scheduler failed
```
Comment (hackergin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java:47]:
```suggestion
     * @throws WorkflowException if closing the related resources of workflow scheduler failed
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * Create a refresh workflow in specific scheduler service for the materialized table, return a
+     * {@link RefreshHandler} instance which can locate the refresh workflow detail information.
+     *
+     * <p>This method supports creating workflow for periodic refresh, as well as workflow for a
+     * one-time refresh only.
+     *
+     * @param createRefreshWorkflow The detail info for create refresh workflow of materialized
+     *     table.
+     * @return The meta info which points to the refresh workflow in scheduler service.
+     * @throws WorkflowException if create refresh workflow failed
```
Comment (hackergin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java:67]:
```suggestion
     * @throws WorkflowException if creating refresh workflow failed
```

Comment (hackergin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java:67]: As I understand it, there may be two possible implementations for the current CreateRefreshWorkflow in the future, corn and one-time. If there is a WorkFlowScheduler that supports only one of these two types, should we throw an unsupported exception at this point?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/WorkflowSchedulerFactoryUtil.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        // no instantiation
+    }
+
+    /**
+     * Attempts to discover the appropriate workflow scheduler factory and creates the instance of
+     * the scheduler. Return null directly if doesn't specify the workflow scheduler in config
+     * because it is optional for materialized table.
+     */
+    public static @Nullable WorkflowScheduler<?> createWorkflowScheduler(
+            Configuration configuration, ClassLoader classLoader) {
```
Comment (hackergin) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/WorkflowSchedulerFactoryUtil.java:60]: Is it possible to pass in a ReadableConfig here? Such as TableConfig? If so, we can directly use the following method to get all options.

```
ConfigOption<Map<String, String>>  workflowSchedulerOptions = new 

ConfigOptions.key(WORKFLOW_SCHEDULER_PREFIX+  "." +  identifier)
                        .mapType()
                        .defaultValue(new HashMap<>());

Map<String, String> options = configuration.get(workflowSchedulerOptions);
```

**Review by hackergin — COMMENTED** [review]
@lsyldliu  Thanks for the contribution,  I left only  two minor comments.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/workflow/WorkflowScheduler.java:67]: I think it is no need, you can the related interface methods of Catalog, it doesn't thrown the  unsupported exception in method signature even though some operations unsupported. Moreover, we have introduced the WorkflowException for WorkflowScheduler, I think we should try to wrap the exception to WorkflowExeception, it is checked Exception, user should sense it.

**Review by lsyldliu — COMMENTED** [review]
@hackergin Thanks for your reviewing, I've addressed your comments.

**Comment by hackergin** [comment]
@lsyldliu  Thanks for the updates and explanations,  LGTM.

**Review by hackergin — APPROVED** [review]

**Pull request merged**